File: week1-c/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long cardNumber = get_long("Number: ");
    long remaining = cardNumber;

    int cardLength = 0;

    int sumCheck = 0;

    int startDigits = 0;
    do
    {
        cardLength++;

        int digit = remaining % 10;

        if (cardLength % 2 == 0) {
            sumCheck = sumCheck + digit * 2 / 10 + digit * 2 % 10;
        } else {
            sumCheck = sumCheck + digit;
        }

        if (remaining > 10 && remaining < 100)
        {
            startDigits = remaining;
        }

        remaining /= 10;
    }
    while (remaining > 0);

    if (sumCheck % 10 != 0 || cardLength < 13 || cardLength > 16)
    {
        printf("INVALID\n");
    } else if (cardLength == 15 && (startDigits == 34 || startDigits == 37))
    {
        printf("AMEX\n");
    } else if (cardLength == 16 && (startDigits >= 51 && startDigits <= 55))
    {
        printf("MASTERCARD\n");
    } else if ((cardLength == 13 || cardLength == 16) && startDigits / 10 == 4)
    {
        printf("VISA\n");
    } else
    {
        printf("INVALID\n");
    }
}

File: week6-python/credit/credit.py
import re

def main():
    card_input = input("Number: ").strip()

    if not re.fullmatch(r"\d{13,16}", card_input):
        print("INVALID")
        return

    card_number = int(card_input)
    remaining = card_number
    card_length = 0
    sum_check = 0
    start_digits = 0

    while remaining > 0:
        card_length += 1
        digit = remaining % 10

        if card_length % 2 == 0:
            doubled = digit * 2
            sum_check += doubled // 10 + doubled % 10
        else:
            sum_check += digit

        if 10 < remaining < 100:
            start_digits = remaining

        remaining //= 10

    if sum_check % 10 != 0:
        print("INVALID")
    elif card_length == 15 and (start_digits == 34 or start_digits == 37):
        print("AMEX")
    elif card_length == 16 and (51 <= start_digits <= 55):
        print("MASTERCARD")
    elif (card_length == 13 or card_length == 16) and start_digits // 10 == 4:
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
